mlflow/entities/run_data.py

```diff
@@ -2,66 +2,69 @@
 from mlflow.entities.metric import Metric
 from mlflow.entities.param import Param
 from mlflow.entities.run_tag import RunTag
-from mlflow.protos.service_pb2 import RunData as ProtoRunData
+from mlflow.protos.service_pb2 import RunData as ProtoRunData, Param as ProtoParam,\
+    RunTag as ProtoRunTag
 
 
 class RunData(_MLflowObject):
     """
     Run data (metrics and parameters).
     """
     def __init__(self, metrics=None, params=None, tags=None):
-        self._metrics = []
-        self._params = []
-        self._tags = []
-        if metrics is not None:
-            for m in metrics:
-                self._add_metric(m)
-        if params is not None:
-            for p in params:
-                self._add_param(p)
-        if tags is not None:
-            for t in tags:
-                self._add_tag(t)
+        """
+        Construct a new :py:class:`mlflow.entities.RunData` instance.
+        :param metrics: List of :py:class:`mlflow.entities.Metric`.
+        :param params: List of :py:class:`mlflow.entities.Param`.
+        :param tags: List of :py:class:`mlflow.entities.RunTag`.
+        """
+        # Maintain the original list of metrics so that we can easily convert it back to
+        # protobuf
+        self._metric_objs = metrics or []
+        self._metrics = {metric.key: metric.value for metric in self._metric_objs}
+        self._params = {param.key: param.value for param in (params or [])}
+        self._tags = {tag.key: tag.value for tag in (tags or [])}
 
     @property
     def metrics(self):
-        """List of :py:class:`mlflow.entities.Metric` for the current run."""
+        """
+        Dictionary of string key -> metric value for the current run.
+        For each metric key, the maximum metric value at the maximum timestamp is returned.
```

Comment:
nit: I would maybe rephrase that to make it clear that we do not expect to have ties. e.g. something like : ... the metric value with the latest timestamp is returned. In case there are multiple values with the same latest timestamp, the maximum of these values is returned.

Just a suggestion though.